## go 的错误处理

Reference

- https://go.dev/blog/error-handling-and-go

接口定义：

```go
type error interface {
    Error() string
}
```

实现错误：

```go
type SyntaxError struct {
    msg    string // description of error
    Offset int64  // error occurred after reading Offset bytes
}

func (e *SyntaxError) Error() string { return e.msg }
```

拓展error接口：

```go
package net

type Error interface {
    error
    Timeout() bool   // Is the error a timeout?
    Temporary() bool // Is the error temporary?
}
```

通过类型断言来更加准确识别错误

```go
if nerr, ok := err.(net.Error); ok && nerr.Temporary() {
    time.Sleep(1e9)
    continue
}
if err != nil {
    log.Fatal(err)
}
```

errors 接口返回值问题

概念：接口是一种由两部分组成的数据结构：一个是类型信息，另一个是数据值。当你将一个具体类型赋值给接口时，接口会存储该类型的信息和对应的值。要使接口等于 nil，接口的类型信息和数据值都必须是 nil

```go
package main

import (
    "fmt"
)

type MyError struct {
    Message string
}

func (e *MyError) Error() string {
    return e.Message
}

func returnNilError() error {
    var err *MyError = nil
    return err
}

func main() {
    err := returnNilError()
    if err != nil {
        fmt.Println("Error is not nil")
    } else {
        fmt.Println("Error is nil")
    }
}


// output: Error is not nil
/*
在 returnNilError 函数中，err 被声明为 *MyError 类型的 nil 值。
当 err 被返回时，它被赋值给一个 error 接口变量。在这个接口中，类型信息是 *MyError，而值是 nil。
因此，当你在 main 函数中检查 err != nil 时，结果为 true，因为接口的类型信息部分不为 nil。
*/

// 正确处理
func returnNilError() error {
    var err *MyError = nil
    if err == nil {
        return nil
    }
    return err
}
```





